File: frontend/src/services/assignmentService.ts
import { api } from "@/lib/api";

import {
  Assignment,
  CreateAssignmentRequest,
  UpdateAssignmentRequest,
} from "@/types/assignment";

class AssignmentService {
  async getAssignments(): Promise<Assignment[]> {
    return api<Assignment[]>("/assignments/");
  }

  async getAssignment(id: string): Promise<Assignment> {
    return api<Assignment>(`/assignments/${id}`);
  }

  async createAssignment(
    data: CreateAssignmentRequest
  ): Promise<Assignment[]> {
    return api<Assignment[]>("/assignments/", {
      method: "POST",
      body: data,
    });
  }

  async updateAssignment(
    id: string,
    data: UpdateAssignmentRequest
  ): Promise<Assignment> {
    return api<Assignment>(`/assignments/${id}`, {
      method: "PUT",
      body: data,
    });
  }

  async deleteAssignment(
    id: string
  ): Promise<{ message: string }> {
    return api<{ message: string }>(
      `/assignments/${id}`,
      {
        method: "DELETE",
      }
    );
  }
}

export const assignmentService = new AssignmentService();

File: backend/app/services/assignment_service.py
from fastapi import HTTPException
from app.database.supabase import supabase
from app.schemas.assignment import AssignmentCreate, AssignmentUpdate


class AssignmentService:

    @staticmethod
    def create_assignment(assignment: AssignmentCreate):
        try:
            data = assignment.model_dump(mode="json")

            response = (
                supabase
                .table("assignments")
                .insert(data)
                .execute()
            )

            return response.data

        except Exception as e:
            raise HTTPException(
                status_code=500,
                detail=f"Failed to create assignment: {str(e)}"
            )

    @staticmethod
    def get_all_assignments():

        try:
            response = (
                supabase
                .table("assignments")
                .select("*")
                .order("due_date")
                .execute()
            )

            return response.data

        except Exception as e:
            raise HTTPException(
                status_code=500,
                detail=str(e)
            )

    @staticmethod
    def get_assignment(assignment_id: str):

        try:
            response = (
                supabase
                .table("assignments")
                .select("*")
                .eq("id", assignment_id)
                .execute()
            )

            if not response.data:
                raise HTTPException(
                    status_code=404,
                    detail="Assignment not found"
                )

            return response.data[0]

        except HTTPException:
            raise

        except Exception as e:
            raise HTTPException(
                status_code=500,
                detail=str(e)
            )

    @staticmethod
    def update_assignment(
        assignment_id: str,
        assignment: AssignmentUpdate
    ):

        try:

            data = assignment.model_dump(
                exclude_none=True,
                mode="json"
            )

            response = (
                supabase
                .table("assignments")
                .update(data)
                .eq("id", assignment_id)
                .execute()
            )

            if not response.data:
                raise HTTPException(
                    status_code=404,
                    detail="Assignment not found"
                )

            return response.data[0]

        except HTTPException:
            raise

        except Exception as e:
            raise HTTPException(
                status_code=500,
                detail=str(e)
            )

    @staticmethod
    def delete_assignment(assignment_id: str):

        try:

            response = (
                supabase
                .table("assignments")
                .delete()
                .eq("id", assignment_id)
                .execute()
            )

            if not response.data:
                raise HTTPException(
                    status_code=404,
                    detail="Assignment not found"
                )

            return {
                "message": "Assignment deleted successfully"
            }

        except HTTPException:
            raise

        except Exception as e:
            raise HTTPException(
                status_code=500,
                detail=str(e)
            )